Switch database to postgresql

carlosmunoz opened the pull request:
Switch to postgresql as the underlying datastore.

Includes:
- A baseline script for creating the database structure from scratch.
- A script to migrate an existing Zanata database (3.7.1+) to postgresql
- Changes to the JPA entities to be compatible with postgresql

**standalone.xml changes**

The only change necessary in standalone.xml is the inclusion of the datasource for Zanata like this as an example:

``` xml
             <datasource jndi-name="java:jboss/datasources/zanataDatasource" 
                    pool-name="zanataDatasource" enabled="true" use-ccm="true">
                    <connection-url>jdbc:postgresql://localhost:5432/zanata</connection-url>
                    <driver-class>org.postgresql.Driver</driver-class>
                    <driver>postgresql-jdbc.jar</driver>
                    <pool>
                        <min-pool-size>0</min-pool-size>
                        <max-pool-size>20</max-pool-size>
                        <flush-strategy>FailingConnectionOnly</flush-strategy>
                    </pool>
                    <security>
                        <user-name>user</user-name>
                        <password>password</password>
                    </security>
                    <statement>
                        <track-statements>NOWARN</track-statements>
                    </statement>
                </datasource>
```

It might also be necessary to add the postgresql jdbc driver to the `deployments` directory in Jboss.

Still needed:
- [x] New base line script generated from the groovy script.
- [x] Triggers on the baseline script
- [x] What to do with changelogs?
- [x] Postgresql is much stricter when it comes to `GROUP BY` statements, so some HQL queries are failing because of this.
- [x] Make sure methods annotated with `@NativeQuery` are migrated to postgresql-compatible SQL.
- [x] Documentation on migration procedures
- [x] Switch functional tests to postgresql

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zanata/zanata-server/938)

<!-- Reviewable:end -->

Commit:
Add a postgresql baseline.

Also change generation type on the ModelEntity base to suit postgresql's generators.

Commit:
Change smallint and bits to not null

Commit:
Make JPA keys Identity generated.

Commit:
Add a script able to create postgresql 'compatible' dumps.

Commit:
Add migration script from mysql to postgresql

carlosmunoz committed:
Add triggers to postgresql baseline.

Also remove unnecessary data migration script.

carlosmunoz committed:
Adapt native queries to postgresql.

djansen-redhat commented:
I tried starting a server and got "PSQLException: ERROR: relation "public.HCommunity_Member" does not exist"

djansen-redhat commented on zanata-war/etc/scripts/Mysql2Postgresql.groovy:
```diff
@@ -0,0 +1,396 @@
+    def options = cli.parse(args)
+
+    def cliArgs = options.arguments()
+    if(options.help) {
```
Suggestion: options.help || args.count == 0
If the user specified no arguments, defaulting to -h would be nice

carlosmunoz replied:
Feel free to add a commit :smile:

djansen-redhat committed:
fix(postgres-script): Show usage on no options given

djansen-redhat committed:
fix(postgres-script): Don't print null or exceptions

Commit:
Add classes to take care of native SQL fragments.

Commit:
Add classes to take care of native SQL fragments.

Commit:
Use native query helpers where necessary.

Also change hql queries to be compatible with Postgresql.

Commit:
Add custom Postgresql dialect

Some things don't work quite right with the out-of-the-box dialect.
For example 'true' and 'false' are not considered keywords.

Commit:
Modify liquibase changesets for Postgresql.

Add validation changesets that check for the right state of the database
before an upgrade to Postgresql.
Add a baseline sql script for Postgresql.

Commit:
Modify functional tests so it doesn't expect mysql

Remove plugins that start an embedded mysql instance.
It now expects a Postgresql database already running.

Commit:
Start a docker postgresql container for FTs.

Requires the docker daemon service to be
running, and the preferably the postgres
docker image to be locally available.

Commit:
Downgrade postgresql dependency for jdk1.7 compatibility.

Correct checkstyle violations.
Make docker run with sudo.

Commit:
Remove pg docker container after functional tests

Commit:
Fix integration tests.

Commit:
Ensure liquibase doesn't quote table or column names.

Also give some time to the container database to
come up before running tests.

Commit:
Remove mentions to unused dependencies

Commit:
Use the custom H2 Dialect with Arquillian

To ensure that booleans are treated properly.

Commit:
Fix problem with Postgresql trigger.

Commit:
Fix failing functional tests

carlosmunoz commented:
Passed all tests in build 4489
(Not showing due to problems with the Github api)

Commit:
Add documentation for the database migration

carlosmunoz commented:
@zanata-jenkins please retest this

carlosmunoz commented:
@zanata-jenkins retest this please

seanf committed:
Modify migration script to handle production database

Add hashbang to make script executable.
Handle illegal null characters in strings.
Use mysql streaming resultsets to conserve memory.

djansen-redhat commented on zanata-war/etc/scripts/Mysql2Postgresql.groovy:
```diff
@@ -0,0 +1,457 @@
+    cli.psqluser(args:1, argName: 'psqluser', 'User for the portgresql database (default: "")')
+    cli.psqlpassword(args:1, argName: 'psqlpassword', 'Password for the postgresql database (default: "")')
+
+    cli.script(args:1, argName: 'script', 'Script file with the PostgreSQL DDL')
```
Should say something like "Output a script file with...". It's a bit unclear as to what it does.

djansen-redhat commented on zanata-war/etc/scripts/Mysql2Postgresql.groovy:
```diff
@@ -0,0 +1,457 @@
+    mysqldb.resultSetConcurrency = ResultSet.CONCUR_READ_ONLY
+    mysqldb.withStatement{stmt -> stmt.fetchSize = Integer.MIN_VALUE }
+
+    if(generateDDL) {
```
if (!generateDDL) ?

carlosmunoz replied:
Just wanted to keep it consistent as in all other places I wrote an 'if(generateDDL) {} else {}' structure. Makes it more readable imo

carlosmunoz replied:
Just keeping it consistent with all other places where the code "asks the same question". Easier to read imo

djansen-redhat commented on zanata-war/etc/scripts/Mysql2Postgresql.groovy:
```diff
@@ -0,0 +1,457 @@
+            fkName = "${fkName}_${fkTableName}"
+        }
+
+//        def fkName = foreignKeys.getString(12)
```
Should this be here?

seanf committed:
Make migration script more user-friendly

Check for outdated database before starting.
Add progress bar for each table.

Commit:
Minor cosmetic changes to migration script

seanf committed:
Replace illegal NUL with space character

djansen-redhat commented:
@carlosmunoz Probably should update zanata-overlay too?

carlosmunoz commented:
Yes. Good catch.

Carlos A. Munoz
Software Engineering Supervisor
Globalization
Red Hat

On Fri, Sep 25, 2015 at 10:41 AM, Damian Jansen <notifications@github.com>
wrote:

> @carlosmunoz <https://github.com/carlosmunoz> Probably should update
> zanata-overlay too?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/938#issuecomment-143090449>.
>

djansen-redhat commented:
- [ ] In testing, Glossary uploads were shown to take 5-10 seconds per database entry, i.e.
```
2015-09-25T00:24:11,323Z INFO  [org.zanata.model.HGlossaryTerm] (default task-48) persist entity: HGlossaryTerm(content=for the British pound, a unit of currency., comments=[], glossaryEntry=HGlossaryEntry(sourceRef=, srcLocale=HLocale(localeId=en-US)), locale=HLocale(localeId=en-US))
```
would take 5-10s to complete

seanf committed:
Move postgres baseline back to 3.7.x

seanf committed:
Generate baseline from a fresh database

seanf committed:
Put back post-3.7 changeset which was removed

seanf commented on zanata-war/etc/scripts/Mysql2Postgresql.groovy:
```diff
@@ -0,0 +1,479 @@
+        def size = columns.getInt(7)
+        def isNullable = "YES".equalsIgnoreCase( columns.getString(18) ) ? true : false
+        def isAutoIncrement = "YES".equalsIgnoreCase( columns.getString(23) ) ? true : false
+        def isGenerated = "YES".equalsIgnoreCase( columns.getString(24) ) ? true : false
```
@carlosmunoz What are all these magic numbers?

carlosmunoz replied:
Those :sparkles:  magic  numbers :sparkles: come from:
http://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getColumns(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String)

seanf replied:
Can we refer to them by name instead of position? Failing that, we should
at least define some constants with meaningful names.

On 25 September 2015 at 16:28, Carlos Munoz <notifications@github.com>
wrote:

> In zanata-war/etc/scripts/Mysql2Postgresql.groovy
> <https://github.com/zanata/zanata-server/pull/938#discussion_r40402686>:
>
> > +        postgresdb.execute(statement)
> > +    }
> > +}
> > +
> > +def createTable(String tableName, ResultSet columns) {
> > +    println "Creating table $tableName"
> > +    StringBuilder createTableSql = new StringBuilder("CREATE TABLE $tableName ( ")
> > +    boolean firstCol = true
> > +    while(columns.next()) {
> > +        def colName = columns.getString(4)
> > +        def dataType = columns.getInt(5)
> > +        def dataTypeName = columns.getString(6)
> > +        def size = columns.getInt(7)
> > +        def isNullable = "YES".equalsIgnoreCase( columns.getString(18) ) ? true : false
> > +        def isAutoIncrement = "YES".equalsIgnoreCase( columns.getString(23) ) ? true : false
> > +        def isGenerated = "YES".equalsIgnoreCase( columns.getString(24) ) ? true : false
>
> Those [image: :sparkles:] magic numbers [image: :sparkles:] come from:
>
> http://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getColumns(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/938/files#r40402686>.
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat

Commit:
Modify overlay module to build with postgresql

seanf committed:
Use Liquibase to manage postgresql triggers

seanf committed:
Add composite index for glossary terms on postgres

Commit:
Return the new row from the postgresql trigger

seanf committed:
Remove trigger from postgres baseline

seanf commented:
If you want to see the SQL as it executes, try running again with these
entries in standalone*.xml:

            <logger category="org.hibernate.SQL">
                <level name="DEBUG"/>
            </logger>
            <logger category="org.hibernate.type">
                <level name="TRACE"/>
            </logger>

Also make sure you have enabled TRACE logging in the CONSOLE logger, eg:

            <console-handler name="CONSOLE">
                <level name="TRACE"/>
                <formatter>
                    <named-formatter name="COLOR-PATTERN"/>
                </formatter>
            </console-handler>


See http://stackoverflow.com/a/1713464/14379

On 25 September 2015 at 10:45, Damian Jansen <notifications@github.com>
wrote:

>
>    - In testing, Glossary uploads were shown to take 5-10 seconds per
>    database entry, i.e.
>
> 2015-09-25T00:24:11,323Z INFO  [org.zanata.model.HGlossaryTerm] (default task-48) persist entity: HGlossaryTerm(content=for the British pound, a unit of currency., comments=[], glossaryEntry=HGlossaryEntry(sourceRef=, srcLocale=HLocale(localeId=en-US)), locale=HLocale(localeId=en-US))
>
> would take 5-10s to complete
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/938#issuecomment-143090800>.
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat

djansen-redhat commented:
Indexing broke!
[lockedupindexing.txt](https://github.com/zanata/zanata-server/files/3325/lockedupindexing.txt)

seanf committed:
Enable cursor ResultSets for PostgreSQL to save memory

Replace MySQL streaming ResultSets with PostgreSQL cursor ResultSets
See https://jdbc.postgresql.org/documentation/head/query.html#fetchsize-example

seanf commented:
Commit 5ae458c may help somewhat with the memory requirements.  It tells postgres to use cursor-based ResultSets wherever we were using streaming ResultSets in mysql.

If not, I think we need to create more `IndexingStrategy` classes, similar to `HTextFlowTargetIndexingStrategy` but for other tables. See also `StreamingDAO` and its subclasses.

For scalability, we really need to be using streaming/cursor ResultSets for all re-indexing, whether we have mysql or postgres underneath.

@djansen-redhat if you get a chance, it would be good to know if mysql can still reindex that big a database without running out of memory.

seanf committed:
Reduce batch size for SimpleClassIndexingStrategy

seanf commented:
@djansen-redhat I've reduced the batch size from 5000 to 100 when indexing, which should reduce memory requirements considerably.

seanf committed:
Merge remote-tracking branch 'origin/master' into postgresql

carlosmunoz commented:
This PR is replaced by https://github.com/zanata/zanata-platform/pull/22
